native/src/modules/fetcher/FetcherModule.js

```diff
@@ -1,47 +1,42 @@
 // @flow
 
+import type { EventChannel } from 'redux-saga'
 import NativeFetcherModule, { NativeFetcherModuleEmitter } from './NativeFetcherModule'
+import { eventChannel } from 'redux-saga'
 import { isEmpty } from 'lodash'
 
 export type TargetFilePathsType = { [path: string]: string }
 
 export type FetchResultType = { [path: string]: {| lastUpdate: string, url: string, errorMessage: ?string |} }
 
-export type ProgressCallbackType = (progress: number) => void
-
 class FetcherModule {
-  // TODO NATIVE-264: Correctly handle already fetching
+  // TODO IGAPP-217: Correctly handle already fetching
   static currentlyFetching = false
 
-  async fetchAsync (
-    targetFilePaths: TargetFilePathsType,
-    progress: ProgressCallbackType
-  ): Promise<FetchResultType> {
+  fetchAsync = (targetFilePaths: TargetFilePathsType): [Promise<FetchResultType>, EventChannel<number>] => {
     if (FetcherModule.currentlyFetching) {
       throw new Error('Already fetching!')
     }
-    if (isEmpty(targetFilePaths)) {
-      return {}
-    }
+
     FetcherModule.currentlyFetching = true
 
-    const subscriptions = []
-    subscriptions.push(NativeFetcherModuleEmitter.addListener('progress', progress))
+    const progressChannel = eventChannel<number>(emitter => {
+      const subscription = NativeFetcherModuleEmitter.addListener('progress', emitter)
+      return () => subscription.remove()
+    })
 
-    try {
-      const result = await NativeFetcherModule.fetchAsync(targetFilePaths)
+    const fetchPromise = !isEmpty(targetFilePaths) ? NativeFetcherModule.fetchAsync(targetFilePaths) : Promise.resolve({})
+    fetchPromise.then(result => {
+      progressChannel.close()
+      FetcherModule.currentlyFetching = false
 
       if (!result) {
         // While testing the app I noticed that the FetchResultType of this function was empty. I am absolutely not sure
         // why this happened. As this cause an inconsistent state it is better to throw an error in this case.
         throw new Error('Fetch failed for some reason!')
       }
-
-      return result
-    } finally {
-      subscriptions.forEach(sub => sub.remove())
-      FetcherModule.currentlyFetching = false
-    }
+    })
+    return [fetchPromise, progressChannel]
```

Comment:
Better use an object here with proper names.

Reply:
You can also split this funciton into two. One that returns just the progressChannel. The other function takes the channel and awaits the promise. Not sure about naming though

Reply:
Extracted creationChannel from fetchAsync, channel is opened and closed now independent of fetching itself